### PR TITLE
Integrated custom-urls into analytics

### PR DESCRIPTION
**Pull request by wachterjohannes**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | missing feature of PR #1927 and fixes https://github.com/sulu-io/sulu-standard/issues/594
| Related issues/PRs | none
| License | MIT
| Documentation PR | none

#### What's in this PR?

This PR integrates the custom-urls into the analytics.

#### Why?

After merging that you can manage own analytic keys for the custom-urls.

### CODE REVIEW

**Comment by danrot** [comment]
changelog missing

File: src/Sulu/Component/Util/WildcardUrlUtil.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    /**
+     * Returns regular expression to match given portal-url.
+     *
+     * @param string $portalUrl
+     *
+     * @return string
+     */
+    private static function getRexexp($portalUrl)
```
Comment (danrot) [inline src/Sulu/Component/Util/WildcardUrlUtil.php]: Shouldn't that be `Regex`? Or even `getRegularExpression`?

File: src/Sulu/Component/Webspace/Url/UrlProviderInterface.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Sulu\Component\Webspace\Url;
+
+use Sulu\Component\Webspace\Url;
+use Sulu\Component\Webspace\Webspace;
+
+/**
+ * Provides urls for given webspace.
+ */
+interface UrlProviderInterface
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Url/UrlProviderInterface.php]: How about calling this one `WebspaceUrlProvider`? Would make things clearer to me.

File: src/Sulu/Component/Util/WildcardUrlUtil.php (hunk 1 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Util/WildcardUrlUtil.php]: fixed

File: src/Sulu/Component/Webspace/Url/UrlProviderInterface.php (hunk 2 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Url/UrlProviderInterface.php]: fixed

**Comment by wachterjohannes** [comment]
@danrot when this is green it should be good to merge

File: src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php [hunk 3]
```diff
@@ -31,13 +32,22 @@ class WebspaceSerializeEventSubscriber implements EventSubscriberInterface
...
 
-    public function __construct(WebspaceManagerInterface $webspaceManager, $environment)
-    {
+    public function __construct(
+        WebspaceManagerInterface $webspaceManager,
+        WebspaceUrlProviderInterface $urlProvider,
+        $environment
+    ) {
         $this->webspaceManager = $webspaceManager;
+        $this->urlProvider = $urlProvider;
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php]: Both variables should also be renamed.

File: src/Sulu/Bundle/ContentBundle/Tests/Unit/EventListener/WebspaceSerializeEventSubscriberTest.php [hunk 4]
```diff
@@ -21,14 +21,16 @@
...
 use Sulu\Component\Webspace\PortalInformation;
 use Sulu\Component\Webspace\Url;
+use Sulu\Component\Webspace\Url\WebspaceUrlProviderInterface;
 use Sulu\Component\Webspace\Webspace;
 
 class WebspaceSerializeEventSubscriberTest extends \PHPUnit_Framework_TestCase
 {
     public function testGetSubscribedEvents()
     {
+        $urlProvider = $this->prophesize(WebspaceUrlProviderInterface::class);
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Tests/Unit/EventListener/WebspaceSerializeEventSubscriberTest.php]: The variable here should also be renamed.

File: src/Sulu/Bundle/CoreBundle/DependencyInjection/Compiler/UrlProviderCompilerPass.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+namespace Sulu\Bundle\CoreBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Collects all url provider and combines them in the chain-provider.
+ */
+class UrlProviderCompilerPass implements CompilerPassInterface
```
Comment (danrot) [inline src/Sulu/Bundle/CoreBundle/DependencyInjection/Compiler/UrlProviderCompilerPass.php]: `WebspaceUrlProviderCompilerPass`

File: src/Sulu/Bundle/ContentBundle/Tests/Unit/EventListener/WebspaceSerializeEventSubscriberTest.php (hunk 4 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Tests/Unit/EventListener/WebspaceSerializeEventSubscriberTest.php]: fixed

File: src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php (hunk 3 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php]: fixed

**Comment by wachterjohannes** [comment]
@danrot have rebased the PR to see if it goes green now.

File: src/Sulu/Bundle/CoreBundle/DependencyInjection/Compiler/UrlProviderCompilerPass.php (hunk 5 above)
Comment (danrot) [inline src/Sulu/Bundle/CoreBundle/DependencyInjection/Compiler/UrlProviderCompilerPass.php]: @wachterjohannes You forgot about this one :wink:

**Comment by wachterjohannes** [comment]
@danrot this is green now

**Comment by danrot** [comment]
When directly switching to the analytics tab after creating a custom url without reloading the entire application the newly created custom url is not showing up in the analytics overlay.

**Comment by danrot** [comment]
The analytics keys are not displayed on the website for Custom URLs.

**Comment by wachterjohannes** [comment]
@danrot this PR was introduced before we changed the generation of request attributes last week. should work now

**Pull request merged**